test/regress_mbedtls.c

```diff
@@ -0,0 +1,1007 @@
+	mbedtls_ssl_context *ssl = malloc(sizeof(mbedtls_ssl_context));//SSL_new(get_ssl_ctx());
+	mbedtls_ssl_init(ssl);
+	mbedtls_ssl_setup(ssl, get_mbedtls_config(MBEDTLS_SSL_IS_SERVER));
+
+	type = (enum regress_mbedtls_type)data->setup_data;
+
+	bev = bufferevent_mbedtls_socket_new(
+		data->base, fd, ssl, BUFFEREVENT_SSL_ACCEPTING,
+		BEV_OPT_CLOSE_ON_FREE|BEV_OPT_DEFER_CALLBACKS);
+	tt_assert(bev);
+
+	bufferevent_setcb(bev, respond_to_number, NULL, eventcb,
+	    (void*)(REGRESS_MBEDTLS_SERVER));
+
+	if (type & REGRESS_MBEDTLS_SLEEP) {
+		struct timeval when = { 1, 0 };
+		event_base_once(data->base, -1, EV_TIMEOUT,
+		    acceptcb_deferred, bev, &when);
+		bufferevent_disable(bev, EV_READ|EV_WRITE);
+	} else {
+		bufferevent_enable(bev, EV_READ|EV_WRITE);
+	}
+
+	/* Only accept once, then disable ourself. */
+	evconnlistener_disable(listener);
+
+end:
+	;
+}
+
+static void
+regress_bufferevent_mbedtls_connect(void *arg)
+{
+	struct basic_test_data *data = arg;
+
+	struct event_base *base = data->base;
+
+	struct evconnlistener *listener;
+	struct bufferevent *bev;
+	struct sockaddr_in sin;
+	struct sockaddr_storage ss;
+	ev_socklen_t slen;
+	mbedtls_ssl_context *ssl;
+//	struct rwcount rw = { -1, 0, 0 };
+	enum regress_mbedtls_type type;
+
+	type = (enum regress_mbedtls_type)data->setup_data;
+
+	memset(&sin, 0, sizeof(sin));
+	sin.sin_family = AF_INET;
+	sin.sin_addr.s_addr = htonl(0x7f000001);
+
+	memset(&ss, 0, sizeof(ss));
+	slen = sizeof(ss);
+
+	listener = evconnlistener_new_bind(base, acceptcb, data,
+	    LEV_OPT_CLOSE_ON_FREE|LEV_OPT_REUSEABLE,
+	    -1, (struct sockaddr *)&sin, sizeof(sin));
+
+	tt_assert(listener);
+	tt_assert(evconnlistener_get_fd(listener) >= 0);
+
+	ssl = malloc(sizeof(mbedtls_ssl_context));
+	tt_assert(ssl);
+	mbedtls_ssl_init(ssl);
+	mbedtls_ssl_setup(ssl, get_mbedtls_config(MBEDTLS_SSL_IS_CLIENT));
+
+	bev = bufferevent_mbedtls_socket_new(
+		data->base, -1, ssl,
+		BUFFEREVENT_SSL_CONNECTING,
+		BEV_OPT_CLOSE_ON_FREE|BEV_OPT_DEFER_CALLBACKS);
+	tt_assert(bev);
+
+	bufferevent_setcb(bev, respond_to_number, free_on_cb, eventcb,
+	    (void*)(REGRESS_MBEDTLS_CLIENT));
+
+	tt_assert(getsockname(evconnlistener_get_fd(listener),
+		(struct sockaddr*)&ss, &slen) == 0);
+	tt_assert(slen == sizeof(struct sockaddr_in));
+	tt_int_op(((struct sockaddr*)&ss)->sa_family, ==, AF_INET);
+
+	tt_assert(0 ==
+	    bufferevent_socket_connect(bev, (struct sockaddr*)&ss, slen));
+	/* Possible only when we have fd, since be_mbedtls can and will overwrite
+	 * bio otherwise before */
+	if (type & REGRESS_MBEDTLS_SLEEP) {
+		//TODO abc
+		//BIO *bio;
+
+		//rw.fd = bufferevent_getfd(bev);
+		//bio = BIO_new_rwcount(0);
+		//tt_assert(bio);
+		//BIO_set_data(bio, &rw);
+		//SSL_set_bio(ssl, bio, bio);
+	}
+	evbuffer_add_printf(bufferevent_get_output(bev), "1\n");
+	bufferevent_enable(bev, EV_READ|EV_WRITE);
+
+	event_base_dispatch(base);
+
+//	tt_int_op(rw.read, <=, 100);
+//	tt_int_op(rw.write, <=, 100);
```

Comment:
So, you just copy-pasted regress_ssl.c? And comment asserts that does not passes? I would prefer if regress_ssl.c will be modified to use openssl/mbedtls

Reply:
the relation between `regress_ssl.c` and `regress_mbedtls.c` is like the relation between `bufferevent_mbedtls.c` and `bufferevent_openssl.c`.
they have similar logic, but different structure.
I think it's hard to merge without an unified SSL interface.

Reply:
>the relation between regress_ssl.c and regress_mbedtls.c is like the relation between bufferevent_mbedtls.c and bufferevent_openssl.c.

If for bufferevent_openssl.c I can more or less agree (due to lot's of SSL_* calls), I'm not sure that I can agree with that fact that test should be copy-pasted (since there should not be too much of different, except for bufferevent_openssl* calls, and this does not looks a lot of changes)

Besides you commented some of asserts, which I'm not sure is acceptable.

If you don't have time to make the code generic, I can take a look, just let me know (but that can take a a few weeks until I can find some spare time for this)

Reply:
OK, I will try to merge mbedtls and openssl test into a unified file.
please wait me for some days

Reply:
Sure, thank you!

Reply:
I have reused `regress_ssl.c` in openssl and mbedtls test.
by extracting openssl related code to `regress_openssl.c`.
and write mbedtls related code to `regress_mbedtls.c`.
`regress_openssl.c` and `regress_mbedtls.c` both include `regress_ssl.c`.
and controlling behaviour of `regress_ssl.c` by macro.

It's sound like a little hack. If you have a more graceful way. please tell me

Reply:
Looks way more better, thank you!

Reply:
Maybe you have enough passion to make the bufferevent_openssl.c code generic?  :)

Reply:
Do you mean use this hacky method? OK, I will have a try.

Reply:
But in a separate patch (so we can drop it if will look too ugly)

Reply:
>It's sound like a little hack. If you have a more graceful way. please tell me

Basically you can do the same thing as bufferevent has, i.e. provide structure with ops, constants add to `ssl-compact.h`

i.e.

```c
# openssl_ops.h
struct le_ssl_ops
{
    void (*SSL_free)(SSL *ssl);
};
# provide openssl_ops.c for it
```

And the use in code like in `bufferevent_ssl.c` (renamed from `bufferevent_openssl.c`):

```
#include "openssl_ops.h"

struct le_ssl_ops *ops = openssl_ops;
ops->free(ssl);
```

And for constant's:

```
# ssl-compact.h

#ifdef EVENT__HAVE_OPENSSL
#define LE_SSL_ERROR_WANT_READ SSL_ERROR_WANT_READ
...
#else /* mbedtls */
#define LE_SSL_ERROR_WANT_READ MBEDTLS_ERR_SSL_WANT_READ
#endif   

Does that sounds ok to you in terms of this PR?